tests/nativeapp/factories.py

```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import json
+import os
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Union
+
+import factory
+import yaml
+
+from tests.testing_utils.files_and_dirs import clear_none_values, merge_left
+
+# TODO:
+# - don't like temp_dir - refactor
+# - rewrite a few more sample tests
+# - pdf path array return with local yml
+# - how do we  parametrize multiple configs?
+# - move factories to proper files/directories
+# - write src/dest pair to pdf from project factory
+# - manifest factory
+# - add util to make readme and setup.sql with defaults?
+
+# TODO
+# - snowflake.local.yml support in V1.*
+
+# TODO after POC:
+# - pdf v2
+
+
+class FactoryNoEmptyDict(factory.DictFactory):
+    @classmethod
+    def _create(cls, *args, **kwargs):
+        if len(kwargs) == 0:
+            return None
+        return cls._build(*args, **kwargs)
+
+
+class PackageFactory(FactoryNoEmptyDict):
+    # Package has no required fields
+    # We can throw a warning here for keys that are not in the schema?!
+    pass
+
+
+class ApplicationFactory(FactoryNoEmptyDict):
+    # We can throw a warning here for keys that are not in the schema?!
+    pass
+
+
+class ArtifactFactory(factory.ListFactory):
+    pass
+
+
+class NativeAppFactory(factory.DictFactory):
+
+    name = factory.Faker("word")
+    artifacts = factory.List([], list_factory=ArtifactFactory)
+    package = factory.SubFactory(PackageFactory)
+    application = factory.SubFactory(ApplicationFactory)
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
+        if kwargs["package"] is None:
+            kwargs.pop("package")
+        if kwargs["application"] is None:
+            kwargs.pop("application")
+        return cls._build(model_class, *args, **kwargs)
+
+
+# TODO: can add utils to this class (to get yml parent path etc)
+@dataclass
+class PdfV10Result:
+    def __init__(self, yml: Union[dict, str], path: Path):
+        self.yml = yml
+        self.path = path
+
+
+class PdfV10Factory(factory.DictFactory):
+
+    definition_version = "1"
+    native_app = factory.SubFactory(NativeAppFactory)
+    env = factory.SubFactory(FactoryNoEmptyDict)
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs) -> PdfV10Result:
+        temp_dir = kwargs.pop("temp_dir", os.getcwd())
```

Comment:
not following the use of temp_dir here. This factory shouldn't care about a temp dir, just a target location for the snowflake.yml no?